src/Processors/Formats/Impl/NativeORCBlockInputFormat.cpp

```diff
@@ -0,0 +1,1098 @@
+}
+
+static ColumnWithTypeAndName readColumnWithDateData(
+    const orc::ColumnVectorBatch * orc_column, const orc::Type *, const String & column_name, const DataTypePtr & type_hint)
+{
+    DataTypePtr internal_type;
+    bool check_date_range = false;
+    /// Make result type Date32 when requested type is actually Date32 or when we use schema inference
+    if (!type_hint || (type_hint && isDate32(*type_hint)))
+    {
+        internal_type = std::make_shared<DataTypeDate32>();
+        check_date_range = true;
+    }
+    else
+    {
+        internal_type = std::make_shared<DataTypeInt32>();
+    }
+
+    const auto * orc_int_column = dynamic_cast<const orc::LongVectorBatch *>(orc_column);
+    auto internal_column = internal_type->createColumn();
+    PaddedPODArray<Int32> & column_data = assert_cast<ColumnVector<Int32> &>(*internal_column).getData();
+    column_data.reserve(orc_int_column->numElements);
+
+    for (size_t i = 0; i < orc_int_column->numElements; ++i)
+    {
+        Int32 days_num = static_cast<Int32>(orc_int_column->data[i]);
+        if (check_date_range && (days_num > DATE_LUT_MAX_EXTEND_DAY_NUM || days_num < -DAYNUM_OFFSET_EPOCH))
+            throw Exception(
+                ErrorCodes::VALUE_IS_OUT_OF_RANGE_OF_DATA_TYPE,
+                "Input value {} of a column \"{}\" exceeds the range of type Date32",
+                days_num,
+                column_name);
+
+        column_data.push_back(days_num);
+    }
+
+    return {std::move(internal_column), internal_type, column_name};
+}
+
+static ColumnWithTypeAndName
+readColumnWithTimestampData(const orc::ColumnVectorBatch * orc_column, const orc::Type *, const String & column_name)
+{
+    const auto * orc_ts_column = dynamic_cast<const orc::TimestampVectorBatch *>(orc_column);
+
+    auto internal_type = std::make_shared<DataTypeDateTime64>(9);
+    auto internal_column = internal_type->createColumn();
+    auto & column_data = assert_cast<ColumnDateTime64 &>(*internal_column).getData();
+    column_data.reserve(orc_ts_column->numElements);
+
+    constexpr Int64 multiplier = 1e9L;
+    for (size_t i = 0; i < orc_ts_column->numElements; ++i)
+    {
+        Decimal64 decimal64;
+        decimal64.value = orc_ts_column->data[i] * multiplier + orc_ts_column->nanoseconds[i];
+        column_data.emplace_back(std::move(decimal64));
+    }
+    return {std::move(internal_column), std::move(internal_type), column_name};
+}
+
+static ColumnWithTypeAndName readColumnFromORCColumn(
+    const orc::ColumnVectorBatch * orc_column,
+    const orc::Type * orc_type,
+    const std::string & column_name,
+    bool inside_nullable,
+    bool allow_null_type,
+    bool skip_columns_with_unsupported_types,
+    bool & skipped,
+    DataTypePtr type_hint = nullptr)
+{
+    if (!inside_nullable && (orc_column->hasNulls || (type_hint && type_hint->isNullable()))
+        && (orc_type->getKind() != orc::LIST && orc_type->getKind() != orc::MAP && orc_type->getKind() != orc::STRUCT))
```

Comment:
Currently clickhouse do not support nullable complex types(maybe it will be supported in my another pr: https://github.com/ClickHouse/ClickHouse/pull/53443). Null list/map/struct null values are replaced with empty/default values whether `input_format_null_as_default` is true or not, pls try ut: `tests/queries/0_stateless/00900_null_array_orc_load.sh`. 

`input_format_null_as_default ` is only useful for those columns with type `canBeInsideNullable` = true

Reply:
Ah, we already have this behavior, makes sense then.

(The behavior seems wrong though. `input_format_null_as_default = false` already throws an exception on nulls in non-complex column where CH type is non-Nullable. This seems to be exactly the same situation: the file has null, the CH type is not nullable.)